src/core/net/ip6.cpp

```diff
@@ -977,6 +978,13 @@ Error Ip6::PassToHost(Message           &aMessage,
     // Do not pass IPv6 packets that exceed kMinimalMtu.
     VerifyOrExit(aMessage.GetLength() <= kMinimalMtu, error = kErrorDrop);
 
+    // When the packet is classified as 'receive' or it has a multicast destination we should always forward it to host.
+    // When the packet's source is mesh-local and the destination is not on the host, we should not forward it to host.
+    VerifyOrExit(aReceive || aMessageInfo.GetSockAddr().IsMulticast() ||
+                     !Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetPeerAddr()) ||
+                     Get<ThreadNetif>().GetUnicastAddresses().ContainsMatching(aMessageInfo.GetSockAddr()),
+                 error = kErrorDrop);
+
```

Comment:
Actually thinking more on this, not sure if we need to check if dest is unicast address of thread netif, since this is already checked when `aReceive` is determined here:
https://github.com/openthread/openthread/blob/df010bcbe5d35af71513662fa6d25e937ace3263/src/core/net/ip6.cpp#L1171-L1179

So if destination is an addressed to `ThreadNeif`, `aReceive` would be set already.

I think we can replace this check with the following
```c++
    // If the sender used mesh-local address as source, do not pass to
    // host unless this message is intended for this device itself.

    if (Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetPeerAddr()))
    {
        VerifyOrExit(aReceive, error = kErrorDrop);
    }
```

Does this sound reasonable?

Reply:
Thanks! This is much more readable.